File: solutions/0313. Super Ugly Number/0313-2.cpp
struct UglyNum {
  int prime;
  int index;   // point to the next index of uglyNums
  long value;  // prime * uglyNums[index]
  UglyNum(int prime, int index, long value)
      : prime(prime), index(index), value(value) {}
};

class Solution {
 public:
  int nthSuperUglyNumber(int n, vector<int>& primes) {
    auto compare = [&](const UglyNum& a, const UglyNum& b) {
      return a.value > b.value;
    };
    priority_queue<UglyNum, vector<UglyNum>, decltype(compare)> pq(compare);
    vector<int> uglyNums{1};

    for (const int prime : primes)
      pq.emplace(prime, 1, prime * uglyNums[0]);

    while (uglyNums.size() < n) {
      uglyNums.push_back(pq.top().value);
      while (pq.top().value == uglyNums.back()) {
        const auto [prime, index, _] = pq.top();
        pq.pop();
        pq.emplace(prime, index + 1, prime * (long)uglyNums[index]);
      }
    }

    return uglyNums.back();
  }
};

File: solutions/0313. Super Ugly Number/0313-2.py
class UglyNum:
  def __init__(self, prime: int, index: int, value: int):
    self.prime = prime
    self.index = index  # poto the next index of uglyNums
    self.value = value  # prime * uglyNums[index]


class Solution:
  def nthSuperUglyNumber(self, n: int, primes: List[int]) -> int:
    pq = []  # (value, prime, index)
    uglyNums = [1]

    for prime in primes:
      heapq.heappush(pq, (prime * uglyNums[0], prime, 1))

    while len(uglyNums) < n:
      uglyNums.append(pq[0][0])
      while pq[0][0] == uglyNums[-1]:
        _, prime, index = heapq.heappop(pq)
        heapq.heappush(pq, (prime * uglyNums[index], prime, index + 1))

    return uglyNums[-1]
